compiler/optimizer/demand.go

```diff
@@ -0,0 +1,177 @@
+package optimizer
+
+import (
+	//"fmt"
+	"github.com/brimdata/zed/compiler/ast/dag"
+)
+
+type Demand interface {
+	isDemand()
+}
+
+// TODO Think about normal form for Demand.
+
+// A nil Demand means 'demand nothing'.
```

Comment:
Maybe it would be safer to have an explicit `DemandNone` and have the zero value for `Demand` just be an error?

Reply:
The Go idiom is return `(nil, err)` is the error indicator so `(nil, nil)` usually has reasonable non-error semantics.  So I think this is ok.